### PR TITLE
Add a lint to warn about un-necessary .into_iter()

### LINKED ISSUE #1094: lint unnecessary `.into_iter()`
`for` loops [implicitly call `.into_iter()`](https://doc.rust-lang.org/std/iter/index.html#for-loops-and-intoiterator) on that which is being iterated over, but if someone didn't know this, they might have written `.into_iter()` themselves ([this is not merely a hypothetical scenario](https://github.com/zackmdavis/Leafline/commit/a4f59709ac6)), which is surely bad style.

Example of code that should get a warning (but does not as of Clippy v. 0.0.79):
```rust
fn main() {
    let my_vec = vec![1, 2, 3, 4];
    for i in my_vec.into_iter() {
        println!("{}", i);
    }
}
```
(could be just `for i in my_vec {`)

I'm guessing the implementation would probably be pretty similar to [that of the](https://github.com/Manishearth/rust-clippy/blob/f1a0ddf31b15d/clippy_lints/src/loops.rs#L497-L511) existing `explicit_iter_loop` lint?

### PR DESCRIPTION
**Pull request by elliottneilclark**
This should close #1094.

### CODE REVIEW

File: clippy_lints/src/loops.rs [hunk 1]
```diff
@@ -577,6 +596,16 @@ fn check_for_loop_arg(cx: &LateContext, pat: &Pat, arg: &Expr, expr: &Expr) {
                                        object,
                                        method_name));
                 }
+            } else if method_name.as_str() == "into_iter" && match_trait_method(cx, arg, &paths::I) {
```
Comment (mcarton) [inline clippy_lints/src/loops.rs]: This should be `INTO_ITERATOR`

File: tests/compile-fail/for_loop.rs [hunk 2]
```diff
@@ -87,7 +87,7 @@ impl Unrelated {
     }
 }
 
-#[deny(needless_range_loop, explicit_iter_loop, iter_next_loop, reverse_range_loop, explicit_counter_loop)]
+#[deny(needless_range_loop, explicit_iter_loop, explicit_into_iter_loop, iter_next_loop, reverse_range_loop, explicit_counter_loop)]
```
Comment (mcarton) [inline tests/compile-fail/for_loop.rs]: We usually just do `#[deny(clippy)]`. I wonder why this file does not use that.

File: tests/compile-fail/for_loop.rs [hunk 3]
```diff
@@ -294,6 +294,10 @@ fn main() {
     for _v in vec.iter() { } //~ERROR it is more idiomatic to loop over `&vec`
     for _v in vec.iter_mut() { } //~ERROR it is more idiomatic to loop over `&mut vec`
 
+
+    let mut _out_vec = vec![1,2,3];
```
Comment (mcarton) [inline tests/compile-fail/for_loop.rs]: Why an leading `_` here?

Comment (elliottneilclark) [inline tests/compile-fail/for_loop.rs]: Not needed. Let me remove it.

File: clippy_lints/src/loops.rs (hunk 1 above)
Comment (elliottneilclark) [inline clippy_lints/src/loops.rs]: Yeah miss commit.

File: tests/compile-fail/for_loop.rs (hunk 2 above)
Comment (elliottneilclark) [inline tests/compile-fail/for_loop.rs]: Don't know for sure. Let me clean that up in a separate pr.

**Comment by mcarton** [comment]
Thanks!

**Pull request merged**